### ISSUE TITLE
SDK for c iothub_client_sample_http able to send message to HUB ,but its not invoking ReceiveMessageCallback(Because of this am not able to receive sent message)

### ISSUE THREAD

**Issue by PRAVEENBADIGER** [Azure/azure-iot-sdk-c#377]
Hi ,

-->Am using Ubuntu 14.0

--> SDK i have used -> 2017-11-17 

# Description of the issue:

I am trying to transfer message from ,client to device(in HUB) and device(in HUB) to client ,using single device connecting string.(using SDK -> iothub_client_sample_http )

Note: i am using client as Ubuntu system  

SDK  iothub_client_sample_http able to send message to cloud(Because i can see incremented message count in azure cloud hub) ,but its not invoking ReceiveMessageCallback(Because of this i was not able to receive message)

Example  for device connecting string which i used:
[redacted]

snap shot shoes message are sent,but the messages are not reverting back
![taminal snapshot](https://user-images.githubusercontent.com/36441456/36146741-9e7a7fb4-10db-11e8-9a49-2fbe3fb5acfb.PNG)

Please help me on this issue .

**Comment by ewertons** [comment]
Hi @PRAVEENBADIGER ,
thanks for your question. We need to confirm you are doing some of the expected things before you proceed.

When you run the http sample above, you will have two modalities exercised: 
1. Device to Cloud messaging (aka telemetry)
2. Cloud to Device messaging (aka c2d)

In the sample you see two callbacks defined:
1. SendConfirmationCallback (for telemetry)
- Gets invoked once the telemetry message sent with IoTHubClient_LL_SendEventAsync is processed (either received by the IoT hub successfully or failed, which is indicated by the result provided)
2. ReceiveMessageCallback (for c2d messaging)
- Gets invoked when a new message is sent from the IoT Hub to the device.

For ReceiveMessageCallback to be invoked, you need to trigger a message to be sent by the IoT Hub to the device. That can be done using DeviceExplorer. 

Question: have you used [DeviceExplorer](https://github.com/Azure/azure-iot-sdk-csharp/tree/master/tools/DeviceExplorer) to send c2d messages to your specific device? 
If not, ReceiveMessageCallback will not be invoked.

**Comment by PRAVEENBADIGER** [comment]
Hi  Ewertons,

Thank you for addressing my issue faster,

First thing is i have not used DeviceExplorer to push message .

I got the clear information how the call back works from above explanation.

Now i have some more doubts to ask you,
Doubt 1:

In  SDK iothub_client_sample_http sample ,(Taking Device 1 as sample)
As i understood the steps are,
1) Assign device1 connecting string  in the code.
2) As per the code we are setting ReceiveMessageCallback by using connecting string  of Device 1 ,So if we receive any message to device 1 ReceiveMessageCallback should invoke  .
3) And we are trying to send messages by using connecting string  of Device 1.at that time SendConfirmationCallback (for telemetry) is invoking .

Now my doubt is ,
We registered a callback for device 1 using  connecting string.
we are sending message to device 1.
why ReceiveMessageCallback is not invoking .(Is it not possible to transfer message to device by using device connecting string ? or explicitly we need to push message to device through DeviceExplorer  )

Doubt 2:

Please tell me this process works or not,
1) In HUB -->i'll add Device 1-->i'll get Device1 connecting string .

2) I configure  SDK1 iothub_client_sample_http sample as only  RECEIVER by Setting ReceiveMessageCallback with  Device1 connecting string.

3) I configure  SDK2 iothub_client_sample_http sample as only TRANSMITTER with  Device1 connecting string and try to sent message .

By the above can i transmit message from SDK2 and Receive message to SDK1 without using any DeviceExplorer ?If Yes,its not working in my system as i logged the problem ,If No,Why? how can i achieve this ?
Note:HTTP protocol is must for me from both the end(TRANSMITTER and RECEIVER).

Thank you

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,

Sorry for listing all the Doubts at the same time ,i am doing this because we work on different time zone .its take long time fro me to reach you back .

As i explained in Doubt 2:

1) From sdk iothub_messaging_sample

IoTHubMessaging_LL_Send(iotHubMessagingHandle, deviceId, messageHandle, sendCompleteCallback, NULL);

The above function IoTHubMessaging_LL_Send taking iotHubMessagingHandle(which is generated from HUB connecting sting), deviceId as inputs ,By this i have achieved sending message to SDK1(Receiver ie:System or controller). 
ie: message going  to --> HUB ,by HUB connecting string  , then its moving to --> device by device id.(its working)

2) From sdk  iothub_client_sample_http

IoTHubClient_LL_SendEventAsync(iotHubClientHandle, messages[iterator].messageHandle, SendConfirmationCallback, &messages[iterator]) != IOTHUB_CLIENT_OK)

The above function IoTHubMessaging_LL_Send taking iotHubClientHandle(which is generated from DEVICE connecting sting),

ie: message going  to --> HUB  and internally routing to --> device, by using device connecting string is this understanding is correct ?(its not working.)

Questions: 
1)What is the difference between above two send messages?
2)Is it not possible to send message to device  using only DEVICE connecting string?
3)What is the importance of device connecting string over Hub connecting string ?
4)I am working on UBUNTU Platform what is the best method to implement, device to cloud and cloud to device messaging using HTTP?

Thank you

**Comment by ewertons** [comment]
Hi @PRAVEENBADIGER ,
I suggest you read a little more about how the Azure IoT messaging works.
Here is a good video: https://channel9.msdn.com/events/Microsoft-Azure/AzureCon-2015/ACON306

There are two clients, the [device client](https://github.com/Azure/azure-iot-sdk-c/tree/master/iothub_client) and the [service client](https://github.com/Azure/azure-iot-sdk-c/tree/master/iothub_service_client).
Telemetry messages sent by the device client go to the IoT hub and internally to something called Event Hub. You can monitor messages landing on the Event Hub using a service client (like Device Explorer).
C2D messages can be sent from a service client (like DeviceExplorer) to the device client.

So as I understand your question, a message (telemetry) sent by the device client with SendAsync will not land back on the device client through the MessageReceivedCallback (which is for C2D messages, # than Telemetry).

To monitor the Telemetry messages sent by the device client you need to use the service client. Unfortunately, we still do not support listening for Telemetry messages using the C service client, but other flavors of the SDK do (C#, Node.Js).

Finally, be aware that a given deviceId shall be used only by one device client endpoint at a time. If you establish more endpoints using the same deviceId, you will impair the previous endpoints currently established.

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,
Thanks for the information,

Can you please explain how can i implement the below using c sdk ,

1) device-device communication through HUB.
2) device-cloud,cloud-device communication.

what all things i need to do additionally ,

Thnak you

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,

Is it possible to Read message(telemetry)  in azure ,sent by the device client .(with any Functional app or with any other method).

Thank  you.

**Comment by ewertons** [comment]
Hi @PRAVEENBADIGER ,
no problem.
Here are the answers:

Can you please explain how can i implement the below using c sdk ,
1. device-device communication through HUB.
A: I would assume you are thinking "device-device" in terms of physical devices, not the logical device as seen by the IoT Hub. To achieve that, you will need two device IDs, e.g. "device1" and "device2". Then you will use the azure-iot-sdk-c device client twice to establish two connections using each of those device IDs. device1 can be set to listen from incoming messages and device2 can send telemetry messages (using *SendAsync). Then you will need a service client (either [C#](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/service/Samples/ConsoleSample/Program.cs) or [node](https://github.com/Azure/azure-iot-sdk-node/blob/master/service/samples/send_c2d_message.js)) to listen for telemetry messages received on the IoT Hub (internally Event Hub) from device2, and send them immediatelly to device1.

2. device-cloud,cloud-device communication.
A: That would be a subset of the explanation above. As an advanced piece of info here, the C service client is capable of sending cloud to device messages, so you wouldn't need either C# or node service clients to achieve that. However, as mentioned before, the C service client is NOT capable yet of receiving device to client (telemetry) messages, so you would need to use one of the other SDK flavors to achieve that.

**Comment by ewertons** [comment]
"Is it possible to Read message(telemetry) in azure ,sent by the device client .(with any Functional app or with any other method)."

Yes, it maps to answer number 2 above.

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,

Thanks for the idea ,

As you told yes am talking about  physical devices (device-device).

About Service client SDK for c : Yes i have used this to transfer messages to device directly ,But the problem i have faced here is, as per my project requirement i need to implement my communication only through HTTP protocol  . Service client SDK for c not supporting to choose the protocol,But device client does(so i am sticking to  device client). 

By your comments i came to know i need to create a service client (either C# or node) which can  listen for telemetry messages received on the IoT Hub (internally Event Hub) from device2, and send them immediately to device1.

As i told i am working on physical devices like raspberry pi,IMX6  controller,I can code only in C .

Questions:
1)  Where i should create Service client, because its possible with only  C# or node .I cannot write in controllers ?
2) Can i create Service client on Cloud directly Is it possible using C# or node SDK?(Can you explain bit) 

Thank you

**Comment by ewertons** [comment]
Hi @PRAVEENBADIGER ,

1. You can run the C# or node client from anywhere that supports it. Unfortunately if your device doesn't support neither, you cannot run the existing Service clients on them for listening for telemetry.
2. There is a solution called Azure Functions that you can use to route messages, but it is not a solution that could scale well if you need to have hundreds of devices.

I am waiting on the official reply from a PM in our team to see if I can give you a better alternative.

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,

I am trying to use functional app,

The process which i followed ,

I have configured Raspberry pi as Device Client using iothub_client_sample_http SDK for C .(which can able to transmit the telemetry messages to cloud).
2)i have deployed functional app on azure as per the instruction given by below links,

https://channel9.msdn.com/Events/TechDaysOnline/MVP-Led-Techdays-Online/Simplifying-IoT-Hub-device-to-device-communication-with-Azure-Functions
https://software.intel.com/en-us/node/731092
Problem ,

I was able to send message to specific device in a hub ,but the the functional app is not getting triggered .

Questions,

1)I have developed Device Client using iothub_client_sample_http SDK for C and ,and built functional app in C# in Azure which takes input from .json file .(will this combination of developing Device client and functional app in different language will work?can normal cSDK device client message  trigger a functional app?or i  need to send message in .json format only?)
2)As per the above question if the answer is YES,Please help me out how can i trigger functional app.
3)Please provide links for functional app codes.

Thank you

**Comment by ewertons** [comment]
Hi @PRAVEENBADIGER ,

You found the solution before me :)
Our PM mentioned Azure Functions as a solution for you. However, be mindful of these two constraints:
A. Azure Functions is a separate paid service from Azure. Whenever you route messages through there you will be billed for that.
B. Azure Functions may not scale well if you have high throughput (more than a few devices sending messages, more than 1 message per second sent)

So here are the answers:
1. Yes, you can combine any language SDKs, the messages are common ground between the different APIs;
2, 3. I would need to find someone versed in Azure Functions to help you here. I'll update this as soon as I find someone.

**Comment by PRAVEENBADIGER** [comment]
Hi Ewertons,

Thanks for you support ,your suggestions are helped me a lot to understand about AZURE HUB  ,looking foreword get support on azure functions.:)

Thank you

**Comment by jspaith** [comment]
Hi @PRAVEENBADIGER 

We're happy to continue supporting IoT specific questions but we're not the Functions experts.  

For basic rampups, there's documentation at https://docs.microsoft.com/en-us/azure/azure-functions/ 

For specific questions once you understand the basics, https://stackoverflow.com/questions/tagged/azure-functions looks pretty active (the 1st page only extends to questions asked yesterday and host have answers already).  There's also https://github.com/Azure/Azure-Functions/issues for problems.

Are there any more IoT concerns you have here?

**Comment by jspaith** [comment]
Since this has been quiet for almost a week, I'm closing it out.  If you have further issues please re-activate this or open a new issue.  Thanks

**Comment by ewertons** [comment]
Following up, one option here is to use event hub client to listen for incoming d2c telemetry and send it as a c2d message to your other device. Please follow up on https://github.com/Azure/azure-iot-sdk-c/issues/461

**Comment by tameraw** [comment]
@PRAVEENBADIGER thank you for your contribution to our open-sourced project!  Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V).
